packages/databyss-editor/lib/clipboardUtils.tsx

```diff
@@ -0,0 +1,476 @@
+    }).after
+
+    // split block at length of selection and get `before`
+    const _secondSplit = splitBlockAtOffset({
+      block: _firstSplit || _block,
+      offset: _selectionLength,
+    }).before
+
+    // if selection is use the first split
+    const _frag = _secondSplit || _firstSplit
+
+    if (_frag) {
+      frag.push({
+        ..._frag,
+        _id: fragmentId(_frag.type, blocks[anchor.index]._id),
+      })
+    }
+  }
+
+  // if selection is more than one block
+  if (anchor.index < focus.index) {
+    blocks.forEach((block: Block, index: number) => {
+      // first block
+      if (index === anchor.index) {
+        const { after: firstBlock } = isAtomicInlineType(block.type)
+          ? { after: block }
+          : splitBlockAtOffset({
+              block: block,
+              offset: anchor.offset,
+            })
+
+        if (firstBlock) {
+          frag.push({
+            ...firstBlock,
+            _id: fragmentId(firstBlock.type, block._id),
+          })
+        }
+      }
+      // get in between frags
+      else if (index > anchor.index && index < focus.index) {
+        const _sliceLength = focus.index - anchor.index
+        //   console.log(_sliceLength)
+        if (_sliceLength > 1) {
+          frag.push({
+            text: block.text,
+            type: block.type,
+            _id: fragmentId(block.type, block._id),
+          })
+        }
+      }
+
+      // last block
+      else if (index === focus.index) {
+        const { before: lastBlock } = splitBlockAtOffset({
+          block: block,
+          offset: focus.offset,
+        })
+
+        if (lastBlock) {
+          frag.push({
+            ...lastBlock,
+            _id: fragmentId(lastBlock.type, block._id),
+          })
+        }
+      }
+    })
+  }
+  // add metadata
+  frag = addBlockUIFields(frag)
+
+  return frag
+}
+
+const mergeText = ({ a, b }: { a: Text; b: Text }): Text => {
+  const mergedTextValue = a.textValue + b.textValue
+
+  const mergedRanges = [
+    ...a.ranges,
+    ...b.ranges.map((r: Range) => ({
+      ...r,
+      offset: r.offset + a.textValue.length,
+    })),
+  ].filter(r => r.length > 0)
+
+  const mergedText = {
+    textValue: mergedTextValue,
+    ranges: mergedRanges,
+  }
+
+  return mergedText
+}
+
+export const insertText = ({
+  block,
+  text,
+  index,
+}: {
+  block: Block
+  text: Text
+  index: number
+}): Block => {
+  const splitBlock = splitBlockAtOffset({ block, offset: index })
+
+  let mergedBlock
+  if (splitBlock.before) {
+    mergedBlock = {
+      text: mergeText({
+        a: splitBlock.before.text,
+        b: text,
+      }),
+      type: 'ENTRY',
+    }
+  }
+
+  if (splitBlock.after) {
+    mergedBlock = {
+      text: mergeText({
+        a: mergedBlock ? mergedBlock.text : text,
+        b: splitBlock.after.text,
+      }),
+      type: 'ENTRY',
+    }
+  }
+  return mergedBlock
+}
+
+export const deleteBlocksAtSelection = ({
+  state,
+  draftState,
```

Comment:
Why do we need `draftState` and `state`. What if we just passed it the draft state and it read/write everything from there?